gptel-context-optimizer: LRU Context Management, one-shot auto add *compilation* buffer, kvcache preloading

aagit opened the pull request:
This PR introduces a LRU (Least Recently Used) context management eliminating the need for manual implementation in `~/.emacs`.

Key features:

1. **Automatic LRU Context**: The LRU context management is now enabled automatically without requiring local configuration
2. **Compilation Buffer Integration**: Adds a configurable option to automatically include the *compilation* buffer at the bottom of the context. This ensures the LLM always has awareness of the most recent compilation issues. Note: this feature is disabled by default as it may not always beneficial.
3. **Relative Path Support**: Obsoletes issue #1113 by keeping relative paths enabled (they are preferable for the LLM too as less is more as far as context is concerned) while fixing the risk of duplication. gptel-context-file-add now does a expanded vs expanded comparison.

karthink commented:
Thanks for the PR @aagit.

> 1. **Automatic LRU Context**: The LRU context management is now enabled
>    automatically without requiring local configuration

This feature looks good, but the implementation needs work.  gptel should not be adding anything to _global_ hooks (like `after-save-hook` and `after-revert-hook`) no matter what.  I would like to avoid adding anything to any buffer's local hooks as well, unless it's a gptel chat buffer.

> 2. **Compilation Buffer Integration**: Adds a configurable option to
>    automatically include the *compilation* buffer at the bottom of the
>    context. This ensures the LLM always has awareness of the most recent
>    compilation issues. Note: this feature is disabled by default as it may not
>    always beneficial.

This is a workflow-specific feature and doesn't fit with gptel's design, so please remove it from this PR.

> 3. **Relative Path Support**: Obsoletes issue #1113 by keeping relative paths
>    enabled (they are preferable for the LLM too as less is more as far as
>    context is concerned) while fixing the risk of
>    duplication. gptel-context-file-add now does a expanded vs expanded
>    comparison.
> 

This change is welcome, but I have some suggestions for the implementation:

1. `gptel-context--file-already-added` returns a matching file, so the name needs to reflect that.  Perhaps `gptel-context--get-path` or `gptel-context--get-file`.
2. This function will fail because `gptel-context` can have file paths and buffers that are not inside lists.  For example, this is a valid value for gptel-context: `(setq gptel-context (list "path1" "path2" (get-buffer "somebuf")))`.  See the documentation of `gptel-context`.
3. There is no need to create a duplicate of `gptel-context` in this function.  Something like the following works better:

   ``` emacs-lisp
     (cl-loop
      with expanded-path = (expand-file-name path)
      for entry in gptel-context
      for source = (or (car-safe entry) entry)
      when (and (stringp source)
                (string= expanded-path (expand-file-name source)))
      return entry)
   ```

   (Or an equivalent while loop.)
4. The message `(message "File \"%s\" added to context." path)` should be sent even if the file is already in the path.  Otherwise there is no feedback when running `gptel-add-file` on a file that is already part of the context.
5. Since we're already testing for membership in `--add-text-file` and `--add-binary-file`, there is no need to use `cl-pushnew`.  Regular `push` is enough.

aagit commented:
> Thanks for the PR @aagit.

You're welcome.

> > 1. **Automatic LRU Context**: The LRU context management is now enabled
> >    automatically without requiring local configuration
> 
> This feature looks good, but the implementation needs work. gptel should not be adding anything to _global_ hooks (like `after-save-hook` and `after-revert-hook`) no matter what. I would like to avoid adding anything to any buffer's local hooks as well, unless it's a gptel chat buffer.

My after-revert-hook is set to quilt-hook. My auto-save-hook is set to gtags-update-hook.


> 
> > 2. **Compilation Buffer Integration**: Adds a configurable option to
> >    automatically include the _compilation_ buffer at the bottom of the
> >    context. This ensures the LLM always has awareness of the most recent
> >    compilation issues. Note: this feature is disabled by default as it may not
> >    always beneficial.
> 
> This is a workflow-specific feature and doesn't fit with gptel's design, so please remove it from this PR.
> 
> > 3. **Relative Path Support**: Obsoletes issue [gptel-context: add file path expansion to prevent duplicates #1113](https://github.com/karthink/gptel/pull/1113) by keeping relative paths
> >    enabled (they are preferable for the LLM too as less is more as far as
> >    context is concerned) while fixing the risk of
> >    duplication. gptel-context-file-add now does a expanded vs expanded
> >    comparison.
> 
> This change is welcome, but I have some suggestions for the implementation:
> 
>     1. `gptel-context--file-already-added` returns a matching file, so the name needs to reflect that.  Perhaps `gptel-context--get-path` or `gptel-context--get-file`.
> 
>     2. This function will fail because `gptel-context` can have file paths and buffers that are not inside lists.  For example, this is a valid value for gptel-context: `(setq gptel-context (list "path1" "path2" (get-buffer "somebuf")))`.  See the documentation of `gptel-context`.
> 
>     3. There is no need to create a duplicate of `gptel-context` in this function.  Something like the following works better:
>        ```emacs-lisp
>          (cl-loop
>           with expanded-path = (expand-file-name path)
>           for entry in gptel-context
>           for source = (or (car-safe entry) entry)
>           when (and (stringp source)
>                     (string= expanded-path (expand-file-name source)))
>           return entry)
>        ```
>        
>        
>            
>              
>            
>        
>              
>            
>        
>            
>          
>        (Or an equivalent while loop.)
> 
>     4. The message `(message "File \"%s\" added to context." path)` should be sent even if the file is already in the path.  Otherwise there is no feedback when running `gptel-add-file` on a file that is already part of the context.
> 
>     5. Since we're already testing for membership in `--add-text-file` and `--add-binary-file`, there is no need to use `cl-pushnew`.  Regular `push` is enough.

aagit commented:
> Also, I think implementing 2 is much simpler than what you're doing right now. Just add the compilation buffer to `gptel-context` _once_, and you're set:
> 
> ```emacs-lisp
> (push (get-buffer "*Compilation*") gptel-context)
> ```
> 
> You shouldn't need to update it again in the Emacs session, as long as the buffer name remains the same.

Yes if you want to just add it, but this is still in the context of the same LRU logic of the first patch: it guarantees at every new build the compilation buffer, if already in context, is removed and re-added at the end, where it's also going to be more effective. The first time you start a build it may invalidate 90% of your context but next time you rebuild it'll be close to zero.

aagit committed:
gptel-context-optimizer: Implement LRU-based context optimization

Files that are recently saved or reverted are moved to the bottom of
the context. This keeps not frequently modified files at the top,
maximizing the KV caching of the context.

Assisted-by: synthmerge

aagit committed:
gptel-context-optimizer: Context kvcache preloading

When a file in context is modified submit a fake gptel-request that
only generated one token in output to keep the kvcache hot, so when a
real gptel-request happens later it'll start generation without delay.

Assisted-by: synthmerge

aagit committed:
gptel-context-optimizer: Add compilation buffer to gptel context

Add the compilation buffer to gptel's context when the compilation
filter runs, ensuring the last compilation output is always available,
but only for the next gptel requests.

This is default disabled and can be enabled with:

    (setq-default gptel-context-enable-compilation-auto-add t)

Assisted-by: synthmerge

aagit committed:
gptel-context-optimizer: Implement LRU-based context optimization

Files that are recently saved or reverted are moved to the bottom of
the context. This keeps not frequently modified files at the top,
maximizing the KV caching of the context.

Assisted-by: synthmerge

aagit committed:
gptel-context-optimizer: Add compilation buffer to gptel context

Add the compilation buffer to gptel's context when the compilation
filter runs, ensuring the last compilation output is always available,
but only for the next gptel requests.

This is default disabled and can be enabled with:

    (setq-default gptel-context-enable-compilation-auto-add t)

Assisted-by: synthmerge

aagit committed:
gptel-context-optimizer: Context kvcache preloading

When a file in context is modified submit a fake gptel-request that
only generated one token in output to keep the kvcache hot, so when a
real gptel-request happens later it'll start generation without delay.

Assisted-by: synthmerge

aagit committed:
gptel-context: Duplicate file prevention

Files with different representations (e.g., relative vs absolute
paths) are now properly deduplicated by pre-expanding filenames before
comparisons.

Assisted-by: synthmerge

aagit committed:
gptel-context: Duplicate file prevention

Files with different representations (e.g., relative vs absolute
paths) are now properly deduplicated by pre-expanding filenames before
comparisons.

Assisted-by: synthmerge

aagit commented:
@karthink ping

karthink commented:
@aagit Haven't lost track of this, just busy!  Will come back to this soon.
